### DISCUSSION TITLE
Consumer shutdown on Commit timeout

### DISCUSSION THREAD

**Discussion by mayurjaiswal9** [confluentinc/parallel-consumer#778] [Q&A]
Hi All,

We are using parallel-consumer-core 0.5.2.8, currently when we face network issues where the consumer is not able to reach the broker, consumer group is shutting down and we see this below error. Once the network is back up we do not see it connecting back again. Should the offset commit timeout be more than the session timeout ? 
I assume this is a misconfiguration, could someone point me to the right values to be set for commit timeout, so that the consumer does not crash ? 

[ERROR] 2024-05-29 10:41:14.468 [pc-control] AbstractParallelEoSStreamProcessor - Error from poll control thread, will attempt controlled shutdown, then rethrow. Error: Timeout waiting for commit response PT30S to request ConsumerOffsetCommitter.CommitRequest(id=e26af99f-df1d-4a43-9d9c-a6c5d9321553, requestedAtMs=1716972064466)

        while (waitingOnCommitResponse) {
            if (attempts > ARBITRARY_RETRY_LIMIT)
                throw new InternalRuntimeException("Too many attempts taking commit responses");

            try {
                log.debug("Waiting on a commit response");
                Duration timeout = AbstractParallelEoSStreamProcessor.DEFAULT_TIMEOUT;
                CommitResponse take = commitResponseQueue.poll(commitTimeout.toMillis(), TimeUnit.MILLISECONDS); // blocks, drain until we find our response
                if (take == null) {
                    throw InternalRuntimeException.msg("Timeout waiting for commit response {} to request {}", timeout, commitRequest);
                }
                waitingOnCommitResponse = take.getRequest().getId() != commitRequest.getId();
            } catch (InterruptedException e) {
                log.debug("Interrupted waiting for commit response", e);
            }
            attempts++;
        }
    }
    

commitTimeout, is injected from offsetCommitTimeout in ParallelConsumerOptions (default value),  

    @Builder.Default
    private final Duration offsetCommitTimeout = Duration.ofSeconds(10);

**Comment by rkolesnev** [comment]
Hi @mayurjaiswal9 - to be honest - I am not 100% sure - there does seem to be an issue with logic handling errors on commit.
For one - in ConsumerOffsetCommitter.java [AsyncCommit block](https://github.com/confluentinc/parallel-consumer/blob/ee18dda3f38ecc1bf592c07ab6583178434d92b7/parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ConsumerOffsetCommitter.java#L95C1-L104C14)
is swallowing exception - I think the idea is to allow for retries / fail on timeout as you are observing.

But I have observed (and there was issue raised previously that went stale) - that the internal Consumer can get stuck and prevent proper shutdown when Kafka cluster becomes unavailable.

Do you have more detailed logs for this? And additional details - what your ParallelConsumer configuration is (i.e. builder / options), KafkaConsumer properties etc...

And maybe it should really be raised as an Issue if you dont mind - could you raise it there?

**Comment by rkolesnev** [comment]
This is the issue that was previously raised - https://github.com/confluentinc/parallel-consumer/issues/597 - may be related.

**Reply by Pigna145** [reply]
That seems to be the same issue (working with Mayur).
To reproduce. Use a docker environment and start a PC and then in the middle of handling messages pause the broker.
Eventually you will see this logging:

>> AbstractParallelEoSStreamProcessor - Error from poll control thread, will attempt controlled shutdown, then rethrow. Error: Timeout waiting for commit response PT30S to request.
And after longer disruptance
>> BrokerPollSystem - Unknown error

(Would expect the PC to go in some kind of check loop to see if it can connect with the broker again)

After that last error the PC is gone (closeOrFailed true). Now the PC has to be restarted. 
In case you are not aware of this situation you can run into an application not consuming anything anymore
Once broker is reachable again it will not join the consumer group immediately and one has to wait until a rebalance is done by the broker.

**Reply by rkolesnev** [reply]
So are you suggesting to keep PC up and try to recover somehow instead of shutting down?
As those kind of errors fall a bit under - oh something external went bad and we dont really know what or why - the retry / check loop would have to be quite specific to some condition that we want to observe as it may never recover.
Basically we would need to classify what is a transient issue related to cluster unavailability - and what is not transient.

**Comment by mayurjaiswal9** [comment]
Thanks for the update.

Adding more details. once we pause the broker(disconnect broker in docker environment)

In the first 10 seconds, we see the below error

io.confluent.parallelconsumer.internal.InternalRuntimeException: Timeout waiting for commit response PT30S to request ConsumerOffsetCommitter.CommitRequest(id=075626a0-6219-4314-b968-d1b93577f6f4, requestedAtMs=1718025763663)
	at io.confluent.parallelconsumer.internal.InternalRuntimeException.msg(InternalRuntimeException.java:23) ~[parallel-consumer-core-0.5.2.8.jar:?]
	at io.confluent.parallelconsumer.internal.ConsumerOffsetCommitter.commitAndWait(ConsumerOffsetCommitter.java:154) ~[parallel-consumer-core-0.5.2.8.jar:?]
	at io.confluent.parallelconsumer.internal.ConsumerOffsetCommitter.commit(ConsumerOffsetCommitter.java:74) ~[parallel-consumer-core-0.5.2.8.jar:?]
	at io.confluent.parallelconsumer.internal.BrokerPollSystem.retrieveOffsetsAndCommit(BrokerPollSystem.java:346) ~[parallel-consumer-core-0.5.2.8.jar:?]
	at io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor.commitOffsetsThatAreReady(AbstractParallelEoSStreamProcessor.java:1222) ~[parallel-consumer-core-0.5.2.8.jar:?]
	at io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor.controlLoop(AbstractParallelEoSStreamProcessor.java:777) ~[parallel-consumer-core-0.5.2.8.jar:?]
	at io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor.lambda$supervisorLoop$9(AbstractParallelEoSStreamProcessor.java:732) ~[parallel-consumer-core-0.5.2.8.jar:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:833) [?:?]


After that, the consumer groups went stale after a longer interval with the error message below. 

Unknown error - org.apache.kafka.common.errors.TimeoutException: Timeout of 60000ms expired before successfully committing offsets {decanting-order-commands-0=OffsetAndMetadata{offset=9611, leaderEpoch=null, metadata='bgAA'}}

**KafkaConsumer.java**

    @Override
    public void commitSync(final Map<TopicPartition, OffsetAndMetadata> offsets) {
        commitSync(offsets, Duration.ofMillis(defaultApiTimeoutMs));  -> 60000ms
    }
	

**BrokerPollSystem.java** 

        try {
            while (runState != CLOSED) {
                handlePoll();

                maybeDoCommit();

                switch (runState) {
                    case DRAINING -> {
                        doPause();
                    }
                    case CLOSING -> {
                        doClose();
                    }
                }
            }
            log.debug("Broker poller thread finished normally, returning OK (true) to future...");
            return true;
        } catch (Exception e) {
            log.error("Unknown error", e); --> Error thrown here.
            throw e;
        }


Below logic used to subscribe and Poll.

    private void subscribeAndPoll(String perPrefixedTopic) {
        log.debug("Subscribing to topic pattern: '{}'.", perPrefixedTopic);
        try {
            parallelStreamProcessor.subscribe(Pattern.compile(perPrefixedTopic));
        } catch (IllegalStateException e) {
            log.error(
                    "Kafka subscriptions have been updated bypassing Common Event, this is not supported.",
                    e);
            throw e;
        } finally {
            log.debug("Start polling for topic pattern: '{}'.", perPrefixedTopic);
            parallelStreamProcessor.poll(createPollMethod());
        }
    }
	
**Question:** 

Is this an expected behavior ? Will this gracefully shutdown the consumer after the timeout ? 
We want to eventually check the network and try to bring up the consumer. any suggestions here if this is ideal ?

**Comment by rkolesnev** [comment]
Hmm, i will need to test this again - I have seen issues when Parallel Consumer is not able to shutdown the Kafka Consumer that cant connect to Broker Cluster - it is more to do with how Kafka Consumer behaves than Parallel Consumer itself per se - but I am not sure from top of my head what exactly will happen in this case / when Brokers come back online.

I think the underlying Kafka Consumer doesnt actually shut down - but keeps retrying to get back to Kafka Cluster - at the same time Parallel Consumer shuts down on error with the closedOrFailed flag.

We could try to model a shutdown followed by restart cycle instead of just shutdown on error if we can somehow reset underlying Kafka Consumer - or Restart cycle can be implemented in application code though a separate Parallel Consumer state monitoring thread - you can check the reason for failure there as well and decide if it should be restarted or not and reset of Kafka Consumer would be up to application then - provide a fresh Kafka Consumer instance to Parallel Consumer.

We could add a listener or have a hook that returns future that is set on Parallel Consumer shutdown - to make that monitoring easier - if that is a path to take.

From one point - i don't want it to make a default behaviour - that Parallel Consumer restarts on errors by itself - i would rather leave it to application developer to decide if application should shut down as well or be restarted. 

Without any new shutdown hooks for Parallel Consumer that may look like this - in naive / example implementation: 
```
    ParallelStreamProcessor<String, String> parallelConsumer;
    boolean shuttingDown = false;

    void run() {
        ExecutorService monitorThread = Executors.newSingleThreadExecutor();
        monitorThread.submit(() -> {
            while (!shuttingDown) {
                if (parallelConsumer == null && parallelConsumer.isClosedOrFailed()) {
                    //shutdown / reset state for anything that needs cleanup on failure.
                    if(parallelConsumer!=null) {
                        cleanup(parallelConsumer);
                    }
                    
                    //setup Kafka Client and ParallelConsumer instance
                    this.parallelConsumer = prepareParallelConsumer(); 
                    
                    //subscribe and kick off polling.
                    subscribeAndPoll(parallelConsumer);
                } else {
                    //to check periodically. 
                    Thread.sleep(1000);
                }
            }
        });
    }
```

**Comment by CMExile** [comment]
Hey @rkolesnev, I just want to add that my team has also seen this issue happening twice in the past two weeks. 

Some of our Kafka brokers are being restarted for maintenance this week, which is causing a commit to fail, and then the Parallel Consumer shuts down and doesn't recover, even after multiple hours. 

I understand your point about the logic for recovering from these errors should be on the developer's side, but I'll also add a +1 in favour of adding some sort of error handling capabilities to the Parallel Consumer (Even if disabled by default).

**Comment by astubbs** [comment]
@mayurjaiswal9, @Pigna145, @CMExile - two things were happening in your logs, and the maintained fork has fixed the first and made the second a decision you get to make.

**What the message was hiding.** In the consumer-commit modes the broker-poll thread is the only thread that answers commit requests. When the broker went away, `commitSync` timed out on that thread and killed it - your "Unknown error" - and every later commit on the control thread then sat out its timeout and reported "Timeout waiting for commit response PT30S". That message described the waiter, not the cause, and the `PT30S` in it was a constant, not the `offsetCommitTimeout` that actually elapsed. So the shutdown was real, the reason was not in the message, and no configuration of the commit timeout could have changed it. @rkolesnev's reading that error handling on the commit path was the problem was right.

**Fixed in [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0)** (astubbs/parallel-consumer#204, astubbs/parallel-consumer#477):

- When the poll thread dies, every waiter is released at once with the poll thread's own exception as the cause, and later commits fail fast on it instead of each waiting out the timeout.
- A commit that exhausts its budget throws `OffsetCommitBudgetExceededException`, naming the budget that ran out, its value and the option to raise - and it says when `offsetCommitTimeout` was below the consumer's `default.api.timeout.ms`, which is the case where no retry was ever reachable. That is the answer to your configuration question: keep `offsetCommitTimeout` above `default.api.timeout.ms` if you want the client's own retries a chance.
- A dead poll thread now closes the `KafkaConsumer` in every commit mode, so the group gets a LeaveGroup immediately instead of the dead member holding its partitions until `max.poll.interval.ms`.

**What it still does, by design:** the instance terminates, as @rkolesnev said, and `isClosedOrFailed()` plus a monitor thread that builds a fresh instance is still the pattern. The option to do otherwise is an open pull request on the fork, not merged, astubbs/parallel-consumer#352: a `CommitFailureHandler` that returns `SHUT_DOWN` (today's behaviour, the default) or `CONTINUE` with a bounded policy that gives up after consecutive exhaustions or too long without a successful commit - which is the "check loop that may never recover" concern answered with a bound. It covers the sync consumer and transactional commit modes; the asynchronous default mode has no commit budget to exhaust, so it is not covered yet. A health-check surface on the interface for that monitor to read is astubbs/parallel-consumer#226 (open).

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
